src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -276,4 +265,100 @@ private NakadiCursor readCursor(final String subscriptionId, final Timeline time
+            throws MyNakadiRuntimeException1, ServiceUnavailableException, NoSuchSubscriptionException,
+            UnableProcessException, IllegalScopeException {
+        final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
+        validateSubscriptionCursors(subscription, cursors);
+        validateSubscriptionReadScopes(subscription, client);
+
+        final String resetStatusPath = format(PATH_ZK_CURSOR_RESET, subscriptionId);
+        ZKSubscription sessionsListener = null;
+        try {
+            // close subscription connections
+            zkHolder.get().create().withMode(CreateMode.EPHEMERAL).forPath(resetStatusPath);
+
+            final String sessionsPath = format("/nakadi/subscriptions/{0}/sessions", subscriptionId);
+            final ZkSubscriptionClient zkSubClient = new CuratorZkSubscriptionClient(subscriptionId, zkHolder.get());
+            final Object lock = new Object();
+            sessionsListener = zkSubClient.subscribeForSessionListChanges(() -> {
+                try {
+                    // wait until all connections are closed by nakadi
+                    if (zkHolder.get().getChildren().forPath(sessionsPath).isEmpty()) {
+                        storeCursorsInZk(subscriptionId, cursors);
+                        synchronized (lock) {
+                            lock.notify();
+                        }
+                    }
+                } catch (final Exception e) {
+                    throw new MyNakadiRuntimeException1("Unexpected problem occurred when resetting cursor", e);
+                }
+            });
+
+            if (zkHolder.get().getChildren().forPath(sessionsPath).isEmpty()) {
+                storeCursorsInZk(subscriptionId, cursors);
+                return;
+            }
+
+            synchronized (lock) {
+                lock.wait(TimeUnit.SECONDS.toMillis(nakadiSettings.getDefaultCommitTimeoutSeconds()));
+            }
+        } catch (final InterruptedException e) {
+            LOG.info(e.getMessage(), e);
+            Thread.currentThread().interrupt();
+            throw new MyNakadiRuntimeException1("Timeout resetting cursor", e);
+        } catch (final Exception e) {
+            LOG.error(e.getMessage(), e);
+            throw new MyNakadiRuntimeException1("Unexpected problem occurred when resetting cursor", e);
+        } finally {
+            if (sessionsListener != null) {
+                sessionsListener.cancel();
+            }
+            try {
+                zkHolder.get().delete().forPath(resetStatusPath);
+            } catch (final KeeperException.NoNodeException e) {
+                throw new UnableProcessException(
+                        "Impossible to reset cursor. Subscription has not been streamed yet", e);
+            } catch (final Exception e) {
+                throw new MyNakadiRuntimeException1("Unexpected problem occurred when deleting zk node", e);
```

Comment:
Why do you throw this basic exception? It is just a common parent for all our runtime exceptions. It is not supposed to be thrown as it doesn't mean anything by itself.
I think it's better to throw the specific exception which describes the actual problem that happened.